rosidl_typesupport_opendds_cpp/resource/srv__type_support.cpp.em

```diff
@@ -104,24 +113,128 @@ void * create_requester__@(service.namespaced_type.name)(
     DDS::Subscriber_var dds_subscriber,
     allocator_t allocator)
 {
-@# TODO: Implement, considering original code in ffe10f9 or earlier
+  @# TODO: remove next line when ready
+  return nullptr;
+
+  using RequesterType = rosidl_typesupport_opendds_cpp::Requester<
+    @(__dds_request_wrapper_msg_type),
+    @(__dds_response_wrapper_msg_type)
+  >;
+
+  @# Register TypeSupports
+  @(__dds_msg_typesupport_type)RequestWrapperTypeSupport_var tsRequest =
+    new @(__dds_msg_typesupport_type)RequestWrapperTypeSupportImpl;
+
+  if (tsRequest->register_type(dds_participant, "@(__dds_request_wrapper_msg_type)_") != DDS::RETCODE_OK) {
+    RMW_SET_ERROR_MSG("request register_type for requester failed");
+    return NULL;
+  }
+
+  @(__dds_msg_typesupport_type)ResponseWrapperTypeSupport_var tsResponse =
+    new @(__dds_msg_typesupport_type)ResponseWrapperTypeSupportImpl;
+
+  if (tsResponse->register_type(dds_participant, "@(__dds_response_wrapper_msg_type)_") != DDS::RETCODE_OK) {
+    RMW_SET_ERROR_MSG("response register_type for requester failed");
+    return NULL;
+  }
+
+  @# Create Topics
+  CORBA::String_var type_name = tsRequest->get_type_name();
+  DDS::Topic_var request_topic = dds_participant->create_topic(request_topic_str,
+                                                      type_name,
+                                                      TOPIC_QOS_DEFAULT,
+                                                      nullptr,
+                                                      OpenDDS::DCPS::DEFAULT_STATUS_MASK);
+
+  if (!request_topic) {
+    RMW_SET_ERROR_MSG("Request create_topic failed");
+    return nullptr;
+  }
+
+  type_name = tsResponse->get_type_name();
+  DDS::Topic_var response_topic = dds_participant->create_topic(response_topic_str,
+                                        type_name,
+                                        TOPIC_QOS_DEFAULT,
+                                        nullptr,
+                                        OpenDDS::DCPS::DEFAULT_STATUS_MASK);
+
+  if (!response_topic) {
+    RMW_SET_ERROR_MSG("Response create_topic failed");
     return nullptr;
+  }
+
+  auto _allocator = allocator ? allocator : &malloc;
+  rosidl_typesupport_opendds_cpp::RequesterParams requester_params(dds_participant);
+
+  requester_params.publisher(dds_publisher);
+  requester_params.subscriber(dds_subscriber);
+  requester_params.request_topic_name(request_topic_str);
+  requester_params.reply_topic_name(response_topic_str);
+  requester_params.request_topic(request_topic);
+  requester_params.reply_topic(response_topic);
+
+  RequesterType * requester = static_cast<RequesterType *>(_allocator(sizeof(RequesterType)));
+  try {
+    new (requester) RequesterType(requester_params);
+  } catch (...) {
+    RMW_SET_ERROR_MSG("C++ exception during construction of Requester");
+    return nullptr;
```

Comment:
I agree. How can we resolve it here without 'deallocator'?

Reply:
What do the other implementations do?  It's OK to add a deallocator parameter if we need it.

Reply:
Best I can think of is free(requester); - would it work?

Reply:
Got out of synch with Adam's comment - RTI has a memory leak, that's how it got here. I am OK with adding deallocator, if Jiang agrees.

Reply:
No, we can't use free here. That's why I took the allocator parameter out in my previous design suggestion (v2). I was using the rmw default allocator/deallocator to avoid this kind of mismatch. The problem with that approach is that it's not flexible for client code to make their own choices. So we decided to put the allocator parameter back. 

What we can do in this case is to use rcutils_allocator_t with the rmw default allocator/deallocator. That structure contains both allocator and deallocator to keep them consistent with each other. 

Frankly, I don't like the code: auto _allocator = allocator ? allocator : &malloc; and its counterpart for free(). That's very easy for client code to mismatch them. That was the reason I took the allocator parameter out in the first place.

Reply:
We'll create a Trello card to address this issue later.